l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,586 @@
+% The declaration is global. The \meta{bitset var} is initially 0.
+%
+%  Bitsets are implemented as string variables consisting of
+%  \texttt{1}'s and \texttt{0}'s.
+%  The rightmost number is the index position 1, so
+%  the string variable can be viewed directly as the binary number.
+%
+% \end{function}
+%
+% \begin{function}[EXP, pTF,added = 2020-12-14]
+%   { \bitset_if_exist:N, \bitset_if_exist:c  }
+%   \begin{syntax}
+%     \cs{bitset_if_exist_p:N} \meta{bitset var}
+%     \cs{bitset_if_exist:NTF} \meta{bitset var} \Arg{true code} \Arg{false code}%
+%   \end{syntax}
+%   Tests whether the \meta{bitset var} exist.
+% \end{function}
+
+%
+% \section{Setting and unsetting bits}
+%
+% \begin{function}[added = 2020-12-13]
+%   { \bitset_set_true:Nn, \bitset_set_true:cn, \bitset_gset_true:Nn, \bitset_gset_true:cn  }
+%   \begin{syntax}
+%     \cs{bitset_set_true:Nn}   \meta{bitset var}  \Arg{integer expression}\\
+%     \cs{bitset_gset_true:Nn}  \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and sets the bit of the
+% index position it represents to 1. Index position are 1-based and set from
+% right to left (so one can view the bitset as a binary number).
+% If needed the length of the bit vector is enlarged.
+% If the \meta{integer expression} evaluates to a negative number, nothing happens.
+% \end{function}
+%
+% \begin{function}[added = 2020-12-13]
+%   { \bitset_set_false:Nn, \bitset_set_false:cn, \bitset_gset_false:Nn, \bitset_set_false:cn }
+%   \begin{syntax}
+%     \cs{bitset_set_false:Nn}   \meta{bitset var}  \Arg{integer expression}\\
+%     \cs{bitset_gset_false:Nn}  \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and unsets the bit of the
+% index position it represents (sets it to 0).
+% The index is 1-based. If the \meta{integer expression} evaluates to a
+% negative number or is larger than the current length of the bit vector
+% nothing happens. If the leading (left most) bit is unset,
+% zeros are not trimmed but stay in the bit vector.
+% \end{function}
+%
+% \section{Using bitsets}
+%
+% \begin{function}[EXP,added = 2020-12-14]
+%   { \bitset_index:Nn, \bitset_index:cn }
+%   \begin{syntax}
+%     \cs{bitset_index:Nn}   \meta{bitset var}  \Arg{integer expression}
+%   \end{syntax}
+% Evaluates the \meta{integer expression} and output \texttt{1} if the bit is
+% set and \texttt{0} otherwise.
+% \end{function}
+%
+% \begin{function}[EXP,added = 2020-12-13]
+%   {\bitset_to_bin:N, \bitset_to_bin:c}
+%   \begin{syntax}
+%     \cs{bitset_to_bin:N} \meta{bitset var}
+%   \end{syntax}
+% This leaves the current value of the bitset expressed as
+% a binary (string) number in the input stream.
+% If no bit has been set yet, the output is zero.
+% \end{function}
+% \begin{function}[EXP,added = 2020-12-13]
+%   {\bitset_to_arabic:N, \bitset_to_arabic:c}
```

Comment:
I was wondering about the name here. I guess we can't use `int` as that is explicitly limited in size. I thought about `bigint`, or `dec`, but I guess they are no clearer. Not asking for a change, just want to make sure the question is asked.

Reply:
In `l3fp` I made `\fp_to_int:n { 1e20 }` give `100000000000000000000`, so I think `\bitset_to_int:N` would be a better name.

Reply:
Imho the to-functions  are like the `\int_to`- functions: they create a formatted representation. \bitset_to_arabic doesn't create an integer variable or converts the bitset into one. Its output is not like \fp_to_int meant as something that is then usable by the int module.  Instead it prints digits.